_posts/2022-11-03-autoscaling-functions.md

```diff
@@ -0,0 +1,212 @@
+
+For this example we deployed the `bycrypt` function with this configuration:
+
+```yaml
+functions:
+  bcrypt:
+    skip_build: true
+    image: alexellis2/bcrypt:0.1.2
+    labels:
+      com.openfaas.scale.min: 1
+      com.openfaas.scale.max: 10
+      com.openfaas.scale.target: 500
+      com.openfaas.scale.type: cpu
+```
+
+The scaling target is configured in milli-CPU, so 1000 accounts for 1 CPU core. For this example we set the scaling target to 500mi. This means the autoscaler will try to ensure the average CPU load per replica stays around half a cpu core by adding or removing replicas of the function.
+
+We run a similar load test as in the previous examples. Over a period of 4 minutes the load is gradually ramped up from 0 to 120 requests per second.
+
+You can find the test script we used [on GitHub](https://github.com/openfaas/openfaas-autoscaler-tests/blob/main/tests/ramp-cpu.js).
+
+![Grafana dashboard showing more replicas of the bycrypt function are added as the CPU load gradually increases.](/images/2022-scaling-functions/ramp-cpu.png)
+> Grafana dashboard showing more replicas of the bycrypt function are added as the CPU load gradually increases.
+
+## Future work
+
+With OpenFaaS functions can be invoked synchronously, for an instant response to the caller, or asynchronously, using a queue. Queued invocations are ideal for long running request, batch processing and background jobs. One of our customers asked if it would be possible to scale functions functions based on the queue depth. This would mean functions can be scaled proactively to process the queued work as quickly as possible.
+
+The new queueing system based on JetStream that we released recently may make this possible and we have created a poc implementation to test scaling based on queue-depth.
+
+> You can learn more about async invocation and the recent changes we made to the OpenFaaS queuing system in this blog post: [The Next Generation of Queuing: JetStream for OpenFaaS](https://www.openfaas.com/blog/jetstream-for-openfaas/)
+
+<blockquote class="twitter-tweet"><p lang="en" dir="ltr">Running some experiments for scaling <a href="https://twitter.com/openfaas?ref_src=twsrc%5Etfw">@openfaas</a> functions based on queue-depth.<br><br>This enables us to scale functions proactively to process the queued work as quickly as possible. <a href="https://t.co/EZloVtB9sC">pic.twitter.com/EZloVtB9sC</a></p>&mdash; Han Verstraete (@welteki) <a href="https://twitter.com/welteki/status/1583463081258020865?ref_src=twsrc%5Etfw">October 21, 2022</a></blockquote> <script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script> 
+
+## Wrapping up
+
```

Comment:
Can we get some links here for related posts?

https://www.openfaas.com/blog/fan-out-and-back-in-using-functions/
https://www.openfaas.com/blog/limits-and-backpressure/
https://www.openfaas.com/blog/pdf-generation-at-scale-on-kubernetes/